### ISSUE TITLE
Some Qed cannot be removed

### ISSUE THREAD

**Issue by eponier** [coq/coq#15925]
#### Description of the problem

As soon as the async mode starts, checking of the universe constraints of proofs is deferred (if I understand correctly). The `Qed` of those proofs are shown in a special color in Coqide. Since https://github.com/coq/coq/pull/15772, these `Qed` have become non editable.

Minimal example:
```
Lemma A : True. Proof. exact I. Qed.
```
Ask Coqide to navigate to the end of the file. This should trigger the async mode. And `Qed` becomes non editable.

#### Coq Version

coqide 8.15.1

**Comment by eponier** [comment]
This is a Coqide-specific bug.

**Comment by ppedrot** [comment]
I think this is the desired behaviour, though. You have to interrupt Coq to modify code that is outside Proof-Qed blocks (including the Qed statements themselves). What would you have expected instead?

**Comment by eponier** [comment]
I mean it's normal that they are not editable while Coq is active. The bug is that they are still non editable even when Coq has finished processing the file.

**Comment by jfehrle** [comment]
I see the same behavior with 8.13.2.  (I double-checked the version in the About box.)  I wasn't able to build 8.12.  No noticed this before?

I'm reluctant to touch this because I'm a bit uncertain on the details of the intended behavior for the async feature (not just this specific issue).  There's no description of the expected behavior in the doc or elsewhere that I'm available of.  No tests.  If someone comes up with a description that the group agrees with, I'll take another look at this, though probably not right away.

**Comment by eponier** [comment]
In 8.7, the `Qed` is editable but in a buggy manner; when we edit it, the state of CoqIDE does not change and the highligting does not reset, this is really similar to the recent bug https://github.com/coq/coq/issues/15861.

In 8.9, 8.11, 8.12, the `Qed` is not editable. The difference w.r.t. the current situation is that when you selected a whole lemma with its proof and deleted that, it worked, while now, the Qed is really not editable unless you backtrack the Coq buffer (or click on "check universes"), so the whole lemma is removed except the `Qed`. What is really better now, on the other hand, is that copy-paste buffer is not corrupted as it was before when you try to edit a non-editable part. Before, the copy-paster buffer recorded a change to an non-editable part, even if it has no effect in the document, meaning that when you undo it will add a character that was not there in the document before. And sometimes it ended up with completely unusable undo. Now at least, this problem seems to have disappeared.

**Comment by jfehrle** [comment]
@gares The debug output shows that CoqIDE receives an `Incomplete` message for the `Qed`.  The only `Complete` message is generated for the tactic before the `Qed` rather than on the `Qed`.  Should this be fixed in Stm or CoqIDE?  If the former, can you tell me what change to make?  Or I could do this in `Coqops.process_feedback`:

![image](https://user-images.githubusercontent.com/1253341/165863699-dd69c8a4-f09c-4af5-9893-72a1ce7bb7a1.png)


The proof is `Lemma A: True. Proof. exact I. Qed.`

```
[DEBUG] Start eval_call Add ((((" Qed.",-4),(4,false)),30),(1,30))
[DEBUG] End eval_call
[DEBUG] Handling coqtop answer
[DEBUG] Marking  [ -4,  0](   30,   35) Qed. [P]
   :
[DEBUG] Feedback ProcessingIn  on 5
[DEBUG] Marking  [ -4,  0](   30,   35) Qed. [P]
   :
[DEBUG] Feedback Incomplete  on 5
[DEBUG] Marking  [ -4,  0](   30,   35) Qed. [I,P]    (* ADD INCOMPLETE FLAG *)
[DEBUG] Feedback Processed  on 5
[DEBUG] Marking  [ -4,  0](   30,   35) Qed. [I]
   :
[DEBUG] Feedback Processed  on 5
[DEBUG] Marking  [ -4,  0](   30,   35) Qed. [I]
   :
[DEBUG] Feedback Complete  on 4     (* ON 4, SHOULDN'T IT BE ON 5? *)
[DEBUG] Marking  [ -3,  0](   21,   30) exact I. []
```

**Comment by jfehrle** [comment]
Yes, that works.  But it seems like an extra bit of trouble for the user that they may need to deal with _all_ the time.  If I step through the proof sentence by sentence with CTRL-down, the Qed is fully checked and marked in green.  If I use run-to-end or run-to-cursor, the Qed is left incomplete.  If you do this on Windows, you don't see the incomplete marking on Qed because workers are disabled by default.  The "Fully check the document" button should be listed in the CoqIDE section of the doc so it's discoverable--I never knew what that button was supposed to do until now.

Should there be an option that causes the check to be done automatically?

Nonetheless, the Qed should be editable when Coq is idle.

**Comment by gares** [comment]
> Should there be an option that causes the check to be done automatically?

There are limitations which don't make this feasible. The plan is to make this go away in the new LSP thing.

It is documented in the async-proofs section of the manual, I guess.
